vpr/src/place/analytical_placement/AnalyticalSolver.cpp

```diff
@@ -162,21 +211,255 @@ void QPHybridSolver::solve(unsigned iteration, PartialPlacement &p_placement) {
+        
+        auto [maxXId, minXId] = boundNode(node_ids_no_duplicate, p_placement.node_loc_x);
+        auto [maxYId, minYId] = boundNode(node_ids_no_duplicate, p_placement.node_loc_y);
+        if (node_ids_no_duplicate.size() <= 1){
+            continue;
+        }
+        // assign arbitrary node as bound node when they are all equal
+        if (maxXId == minXId) {
+            maxXId = node_ids_no_duplicate[0];
+            minXId = node_ids_no_duplicate[1];
+        }
+        if (maxYId == minYId) {
+            maxYId = node_ids_no_duplicate[0];
+            minYId = node_ids_no_duplicate[1];
+        }
+        auto add_node = [&](size_t first_node_id, size_t second_node_id, unsigned num_nodes, bool is_x){
+            if (!p_placement.is_moveable_node(first_node_id)) {
+                if (!p_placement.is_moveable_node(second_node_id)) {
+                    return;
+                }
+                std::swap(first_node_id, second_node_id);
+            }
+            double epsilon = 1e-6;
+            double dist = 0; 
+            if(is_x)
+                dist = 1.0/std::max(std::abs(p_placement.node_loc_x[first_node_id] - p_placement.node_loc_x[second_node_id]), epsilon);
+            else
+                dist = 1.0/std::max(std::abs(p_placement.node_loc_y[first_node_id] - p_placement.node_loc_y[second_node_id]), epsilon);
+            double w = (2.0/static_cast<double>(num_nodes - 1)) * dist;
+            if (p_placement.is_moveable_node(second_node_id)) {
+                if (is_x) {
+                    tripletList_x.emplace_back(first_node_id, first_node_id, w);
+                    tripletList_x.emplace_back(second_node_id, second_node_id, w);
+                    tripletList_x.emplace_back(first_node_id, second_node_id, -w);
+                    tripletList_x.emplace_back(second_node_id, first_node_id, -w);
+                }else{
+                    tripletList_y.emplace_back(first_node_id, first_node_id, w);
+                    tripletList_y.emplace_back(second_node_id, second_node_id, w);
+                    tripletList_y.emplace_back(first_node_id, second_node_id, -w);
+                    tripletList_y.emplace_back(second_node_id, first_node_id, -w);
+                }
+            } else {
+                if (is_x){
+                    tripletList_x.emplace_back(first_node_id, first_node_id, w);
+                    b_x(first_node_id) += w * p_placement.node_loc_x[second_node_id];
+                }else{
+                    tripletList_y.emplace_back(first_node_id, first_node_id, w);
+                    b_y(first_node_id) += w * p_placement.node_loc_y[second_node_id];
+                }
+            }
+        };
+        for (size_t node_id = 0; node_id < node_ids_no_duplicate.size(); node_id++) {
+            if (node_id != maxXId && node_id != minXId) {
+                add_node(node_id, maxXId, node_ids_no_duplicate.size(), true);
```

Comment:
To prevent confusion, I would recommend using a variable for number of pins here. It is not immediately obvious that the number of nodes in this vector is the same thing. I would also add a comment, so we know where to change when we upgrade the Netlist.